pyfolio/perf_attrib.py

```diff
@@ -612,6 +612,9 @@ def _align_and_warn(returns,
             )
             warnings.warn(warning_msg)
 
+    factor_loadings = factor_loadings.copy()
```

Comment:
This might be very expensive if `factor_loadings` is large. Where are we actually using the dataframe that we depend explicitly on the names of the index levels? In my experience that's relatively rare (I can't actually think of a case offhand where I've depended on index level names).

Reply:
It's in empyrical, when we compute exposures: `ValueError: cannot join with no level specified and no overlapping names`

Reply:
Is there a way for us to do that computation without depending on index level name? In general, I don't think we (or at least I) don't think of index level names as being part of the interface of a function. If we require a particular set of names, then we should be documenting that requirement (and I'd also probably argue that we shouldn't be using `ticker`).

Reply:
Also, it seems like whatever we do for `factor_loadings`, we should be doing for `positions` as well? Is there a reason we're only modifying `factor_loadings` here?

Reply:
Ok, I'll close this PR and open one in empyrical then. For @mmargenot's current alphalens work, should he just change the factor_loadings index name for the time being?

Reply:
We vectorize the positions in pyfolio, and then make this change for positions

Reply:
w/r/t the performance concern, we might be able to avoid that by passing `deep=False` to the `copy` call here, but I think my question still stands about we're assuming that positions has a particular set of level names but not assuming the same for `factor_loadings`.

Reply:
We set the level names for positions when we vectorize them in `perf_attrib._stack_positions`. Only reason the level names matter is to avoid the `ValueError` that Max saw earlier - but like you said I think we can avoid that by doing the computation without depending on the index level name, which would happen in empyrical

Reply:
> For @mmargenot's current alphalens work, should he just change the factor_loadings index name for the time being?

I think that depends on what contract we want to provide here. Do we want PyFolio to require users to have to care about things like the level names of their DataFrames (my guess is no).  If not, then I don't think alphalens should have to be changed. Also, if we make that decision, then I think we still want the test in this PR to ensure that we don't care about user-provided level names.

Reply:
Also, to be clear, "Is there a way for us to do that computation without depending on index level name?" was an actual question.  I'm not sure offhand if there's a straightforward way to do a matrix multiple in pandas on multi-indexed frames with different index level names. So the right fix might very well be for empyrical to be strict and require nicely-formatted input (which in this case would mean requiring that the index level names match), and for PyFolio to do something like what we're doing here (e.g., I think doing a copy with `deep=False` should be fine performance-wise).

Reply:
If I change `compute_exposures` in empyrical to do `groupby(level=0)` rather than `groupby(level='dt').sum()`, perf_attrib will work, but the output of `perf_attrib` will have index names matching the index names of `factor_loadings`.

I think it might be slightly confusing if the outputs of perf_attrib have different index names depending on the input, so I changed copy to use `deep=False` - lmk if you think that's okay.

Reply:
> I think it might be slightly confusing if the outputs of perf_attrib have different index names depending on the input, so I changed copy to use deep=False - lmk if you think that's okay.

So, if empyrical is doing `groupby(level='dt')`, that means that the interface required by empyrical is that the level 0 must be named 'dt'. That looks like it doesn't match our convention in alphalens, where we appear to use the convention of `'date'`. In another internal Quantopian project, I see similar tables with index level names of `'dates'` and `'sid'`. I also see a variety of different usages just within pyfolio (e.g., I see both `date` and `dt`, as well as both `symbol` and `ticker`).

Given our inconsistent usage across projects, I think my overall preference is for `empyrical` to be agnostic to the names of the indices it receives by using positional level lookups. (i.e., using `groupby(level=0)`), as you suggest.

> I think it might be slightly confusing if the outputs of perf_attrib have different index names depending on the input,

Can you give an example where you think this produces a surprising result? Naively, this doesn't seem like that big of a deal to me, but I'm having trouble imagining a specific example.

Reply:
I meant that if `factor_loadings.index.names` is `['asdf', 'asset']`, the exposures/attributed returns dataframes would have an index name of `asdf`, which might be confusing if you expect the output to be consistently formatted

Reply:
That doesn't seem that surprising to me. If I hand you a dataframe with `asdf` as the top-level index name, and you do a groupby computation with that dataframe, I wouldn't be surprised to see `asdf` as the name of the resulting index. If anything, I'd be more surprised to get back a dataframe with an index name that's different from anything I passed in.